### PR TITLE
Deploy fluentbit as a statefulset

### PR DESCRIPTION
**Pull request by wenchajun**
Signed-off-by: chengdehao <dehaocheng@kubesphere.io>

### Which issue(s) this PR fixes:

Fixes #

### Does this PR introduced a user-facing change?

```release-note

```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by wenchajun** [comment]
related https://github.com/fluent/fluent-operator/pull/443

File: pkg/operator/fbstatefulset.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+					Volumes: []corev1.Volume{
+						{
+							Name: "config",
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{
+									SecretName: co.Spec.FluentBitConfigName,
+								},
+							},
+						},
+						{
```
Comment (wanjunlei) [inline pkg/operator/fbstatefulset.go:44]: Does it need to mount the container and system log paths in the Collector mode?

File: apis/fluentbit/v1alpha2/collector_types.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	// Pod volumes to mount into the container's filesystem.
+	VolumesMounts []corev1.VolumeMount `json:"volumesMounts,omitempty"`
+	// Annotations to add to each Fluentbit pod.
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// SecurityContext holds pod-level security attributes and common container settings.
+	SecurityContext *corev1.PodSecurityContext `json:"securityContext,omitempty"`
+	// Host networking is requested for this pod. Use the host's network namespace. If this option is set, the ports that will be used must be specified. Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
+	// Buffer definition
+	BufferVolume *BufferVolume `json:"buffer,omitempty"`
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/collector_types.go:71]: Why need this? This volume is just created but not mounted.

Comment (wenchajun) [reply apis/fluentbit/v1alpha2/collector_types.go:71]: Mounted here
https://github.com/fluent/fluent-operator/pull/484/files#diff-ad79a92078fc8d5c14743ecf902fe8ddcf3ee9d0e79e2e0cc5fd39062f14811bR167

**Comment by benjaminhuo** [comment]
Add FluentBit collector is helpful in scenarios of collecting data through the network including the following input plugins:
- https://docs.fluentbit.io/manual/pipeline/inputs/opentelemetry
- https://docs.fluentbit.io/manual/pipeline/inputs/prometheus-scrape-metrics
- https://docs.fluentbit.io/manual/pipeline/inputs/collectd
- https://docs.fluentbit.io/manual/pipeline/inputs/forward
- https://docs.fluentbit.io/manual/pipeline/inputs/http
- https://docs.fluentbit.io/manual/pipeline/inputs/mqtt
- https://docs.fluentbit.io/manual/pipeline/inputs/nginx
- https://docs.fluentbit.io/manual/pipeline/inputs/statsd
- https://docs.fluentbit.io/manual/pipeline/inputs/syslog
- https://docs.fluentbit.io/manual/pipeline/inputs/syslog
- https://docs.fluentbit.io/manual/pipeline/inputs/tcp

All DaemonSet based fields can be removed:
- ContainerLogRealPath
- PositionDB (Maybe this can also be removed, because there is no position to store in network cases? @wanjunlei what do you think?)

File: apis/fluentbit/v1alpha2/collector_types.go (hunk 2 above)
Comment (wanjunlei) [reply apis/fluentbit/v1alpha2/collector_types.go:71]: What does `BufferVolume` do?

**Comment by wanjunlei** [comment]
> Add FluentBit collector is helpful in scenarios of collecting data through the network including the following input plugins:
> 
> * https://docs.fluentbit.io/manual/pipeline/inputs/opentelemetry
> * https://docs.fluentbit.io/manual/pipeline/inputs/prometheus-scrape-metrics
> * https://docs.fluentbit.io/manual/pipeline/inputs/collectd
> * https://docs.fluentbit.io/manual/pipeline/inputs/forward
> * https://docs.fluentbit.io/manual/pipeline/inputs/http
> * https://docs.fluentbit.io/manual/pipeline/inputs/mqtt
> * https://docs.fluentbit.io/manual/pipeline/inputs/nginx
> * https://docs.fluentbit.io/manual/pipeline/inputs/statsd
> * https://docs.fluentbit.io/manual/pipeline/inputs/syslog
> * https://docs.fluentbit.io/manual/pipeline/inputs/syslog
> * https://docs.fluentbit.io/manual/pipeline/inputs/tcp
> 
> All DaemonSet based fields can be removed:
> 
> * ContainerLogRealPath
> * PositionDB (Maybe this can also be removed, because there is no position to store in network cases? @wanjunlei what do you think?)

Agree.

**Comment by wenchajun** [comment]
fixed

**Comment by benjaminhuo** [comment]
Need to add logic to create a service and add input plugins' port to this service

File: apis/fluentbit/v1alpha2/collector_types.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	HostNetwork bool `json:"hostNetwork,omitempty"`
+	// PVC definition
+	PersistentVolumeClaim *corev1.PersistentVolumeClaim `json:"pvc,omitempty"`
+	// RBACRules represents additional rbac rules which will be applied to the fluent-bit clusterrole.
+	RBACRules []rbacv1.PolicyRule `json:"rbacRules,omitempty"`
+	// By default will build the related service according to the globalinputs definition.
+	DisableService bool `json:"disableService,omitempty"`
+	// The path where buffer chunks are stored. This field would make no effect in memory buffer plugin.
+	// +kubebuilder:validation:Required
+	BufferPath *string `json:"bufferPath,omitempty"`
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/collector_types.go:74]: `kubebuilder:validation:Required` and `omitempty` should not be set simultaneously.

File: controllers/collector_controller.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+	default:
+	}
+
+	return nil
+}
+
+func (r *CollectorReconciler) delete(ctx context.Context, co *fluentbitv1alpha2.Collector) error {
+	var sa corev1.ServiceAccount
+	err := r.Get(ctx, client.ObjectKey{Namespace: co.Namespace, Name: co.Name}, &sa)
```
Comment (wanjunlei) [inline controllers/collector_controller.go:173]: No need to get it before deleting, just delete it directly.

Why does not delete the `clusterrole` and `clusterrolebinding`?

File: controllers/collector_controller.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	if err == nil {
+		if err := r.Delete(ctx, &sa); err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+	} else if !errors.IsNotFound(err) {
+		return err
+	}
+
+	var sts appsv1.StatefulSet
+	err = r.Get(ctx, client.ObjectKey{Namespace: co.Namespace, Name: co.Name}, &sts)
```
Comment (wanjunlei) [inline controllers/collector_controller.go:183]: No need to get it before deleting, just delete it directly.

File: pkg/operator/fbstatefulset.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	return pvc
+}
+
+func FlunetbitBufferMountPath(co fluentbitv1alpha2.Collector) string {
+	bufferPath := co.Spec.BufferPath
+	if bufferPath != nil {
+		if strings.HasPrefix(*bufferPath, "/buffers") {
+			return *bufferPath
+		} else {
+			targetPaths := []string{"/buffers"}
```
Comment (wanjunlei) [inline pkg/operator/fbstatefulset.go:196]: Why do this?

**Review by wanjunlei — DISMISSED** [review]

File: controllers/collector_controller.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+		}
+		return []string{owner.Name}
+	}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&fluentbitv1alpha2.Collector{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&appsv1.StatefulSet{}).
```
Comment (benjaminhuo) [inline controllers/collector_controller.go:221]: The service should also be controlled/owned by the `Collector` CRD

File: controllers/collector_controller.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+	return nil
+}
+
+func (r *CollectorReconciler) delete(ctx context.Context, co *fluentbitv1alpha2.Collector) error {
+	var sa corev1.ServiceAccount
+	if err := r.Delete(ctx, &sa); err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
```
Comment (benjaminhuo) [inline controllers/collector_controller.go:176]: should also add logic the delete the service?

File: pkg/operator/fbstatefulset.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
+package operator
```
Comment (benjaminhuo) [inline pkg/operator/fbstatefulset.go:1]: This file can be renamed to `collector-statefulset`?

File: controllers/collector_controller.go (hunk 8 above)
Comment (wenchajun) [reply controllers/collector_controller.go:176]: fixed

**Review by benjaminhuo — APPROVED** [review]

**Comment by benjaminhuo** [comment]
Don't forget to add docs for this @wenchajun

**Pull request merged**